lib/std/os.zig

```diff
@@ -6281,47 +6283,65 @@ var has_copy_file_range_syscall = std.atomic.Atomic(bool).init(true);
         var off_out_copy = @bitCast(i64, off_out);
 
-        const rc = system.copy_file_range(fd_in, &off_in_copy, fd_out, &off_out_copy, len, flags);
-        switch (system.getErrno(rc)) {
-            .SUCCESS => return @intCast(usize, rc),
-            .BADF => return error.FilesOpenedWithWrongFlags,
-            .FBIG => return error.FileTooBig,
-            .IO => return error.InputOutput,
-            .ISDIR => return error.IsDir,
-            .NOMEM => return error.OutOfMemory,
-            .NOSPC => return error.NoSpaceLeft,
-            .OVERFLOW => return error.Unseekable,
-            .PERM => return error.PermissionDenied,
-            .TXTBSY => return error.SwapFile,
-            // these may not be regular files, try fallback
-            .INVAL => {},
-            // support for cross-filesystem copy added in Linux 5.3, use fallback
-            .XDEV => {},
-            // syscall added in Linux 4.5, use fallback
-            .NOSYS => {
-                has_copy_file_range_syscall.store(false, .Monotonic);
-            },
-            else => |err| return unexpectedErrno(err),
+        while (true) {
+            const rc = system.copy_file_range(fd_in, &off_in_copy, fd_out, &off_out_copy, len, flags);
+            switch (builtin.os.tag) {
+                .linux => {
+                    switch (system.getErrno(rc)) {
+                        .SUCCESS => return @intCast(usize, rc),
+                        .BADF => return error.FilesOpenedWithWrongFlags,
+                        .FBIG => return error.FileTooBig,
+                        .IO => return error.InputOutput,
+                        .ISDIR => return error.IsDir,
+                        .NOMEM => return error.OutOfMemory,
+                        .NOSPC => return error.NoSpaceLeft,
+                        .OVERFLOW => return error.Unseekable,
+                        .PERM => return error.PermissionDenied,
+                        .TXTBSY => return error.SwapFile,
+                        // these may not be regular files, try fallback
+                        .INVAL => break,
+                        // support for cross-filesystem copy added in Linux 5.3, use fallback
+                        .XDEV => break,
+                        // syscall added in Linux 4.5, use fallback
+                        .NOSYS => {
+                            has_copy_file_range_syscall.store(false, .Monotonic);
+                            break;
+                        },
+                        else => |err| return unexpectedErrno(err),
+                    }
+                },
+                .freebsd => {
+                    switch (system.getErrno(rc)) {
+                        .SUCCESS => return @intCast(usize, rc),
+                        .INTR => continue,
+                        .BADF => return error.FilesOpenedWithWrongFlags,
+                        .FBIG => return error.FileTooBig,
+                        .INVAL => return error.RangeOverlap,
```

Comment:
The [Linux manpage](https://man.archlinux.org/man/copy_file_range.2) states that `EINVAL` is returned for overlapping ranges too. Since the original impl. didn't add a specific error for this, I don't think it's worth adding it here. Though I think that was more to do with `EINVAL` being returned for multiple reasons.

Reply:
thanks for review @The-King-of-Toasters . i will check/fix all in a few hours